### DISCUSSION TITLE
Nonempty field take validation error "Missing required value"

### DISCUSSION THREAD

**Discussion by m-2k** [pocketbase/pocketbase#3611] [Q&A] [answered]
Setting up database schema restrictions imposes a limitation that I could not solve using hooks.

I want 3 things at the same time:
1. Be able to use the `create` API method without `status` field
2. **Be able to set constraint on a** `status` **field according to the database schema**
3. Be able to set the default value for the `status` field

<img width="659" alt="Screenshot 2023-10-26 at 18 00 47" src="https://github.com/pocketbase/pocketbase/assets/6705016/7c630396-77fe-4c45-9a65-ccbf4e93df9f">

```go
app.OnModelBeforeCreate("orders").Add(func(e *core.ModelEvent) error {
  rec := e.Model.(*models.Record)
  rec.Set("status", "created")
  return nil
})
```
```js
const _ = await pocketbase.collection('orders').create({})
```
```bash
2023/10/26 17:48:29 status: Missing required value.
```

Any ideas? Thx.

**Comment by ganigeorgiev** [comment]
At the moment record validations happens before the hook handlers (_this may change in the future_).

You can set the default value client-side OR remove the "Nonempty" constraint and set the default value on the server-side:

```go
if rec.GetString("status") == "" {
   rec.Set("status", "created")
}
```

**Reply by m-2k** [reply]
Your check for an empty string is useless since an empty value will be rejected by the server before the `OnModelBeforeCreate` function is called

**Reply by ganigeorgiev** [reply]
That's why I mentioned to remove the "Nonempty" constraint.

**Reply by m-2k** [reply]
if the `Nonempty` constraint is removed, it will also be necessary to write a function for updating `OnModelBeforeUpdate`, so that when updating a record with an empty value, an error in the administrative panel (UI)
That is, it is more like a crutch that needs to be created instead of standard constraint for a database schema column.

It seems to me that the most correct solution would be to add the functionality of setting a [default value](https://www.sqlite.org/lang_createtable.html#the_default_clause) for the database schema, that is, add a `default` field in the administrative panel, what do you think?

**Reply by ganigeorgiev** [reply, marked as answer]
Default values were discussed several times and for now there are no plans for this as it is ambiguous when the default value will have to be applied. My recommendation is to submit the default value with the request client-side or set it manually on the server-side. In the future hooks may get refactored to support before/after validations, but for now other tasks are with higher priority.

**Reply by m-2k** [reply]
My solution without any server-side logic code:
1. Table: `orders`, view: `ordersView` and `API rules` only for both
2. Table columns:
  * `id`: system and auto-generated by server-side
  * `user`: relation to table `users`, public
  * `status`: user only allowed options to set: `created | cancelled`, other (admin) options: `verified | accepted | refused | work | finished`; user can change field only from `created | cancelled | verified`, must be equal `created` for `create` API request
  * `payload`: user can modify and read this field
  * `comment`: user can modify and read this field
  * `remark`: hidden field for user

3. Table
<img width="633" alt="Screenshot 2023-10-26 at 20 49 11" src="https://github.com/pocketbase/pocketbase/assets/6705016/a0f8cd37-cb6e-404b-88cc-47ab77014fc5">

4. Table API rules (only `create` and `update` requests allowed for user)
<img width="633" alt="Screenshot 2023-10-26 at 20 49 35" src="https://github.com/pocketbase/pocketbase/assets/6705016/d2bb14d4-88a1-41aa-a119-1cdb189cbe13">

5. View
<img width="632" alt="Screenshot 2023-10-26 at 20 49 57" src="https://github.com/pocketbase/pocketbase/assets/6705016/f12e3312-6591-445d-b175-9a76656cf215">

6. View API rules (user can read only self records)
<img width="633" alt="Screenshot 2023-10-26 at 20 50 09" src="https://github.com/pocketbase/pocketbase/assets/6705016/ca674217-ace1-4fa8-ace7-91c53499a073">
